src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -110,6 +116,24 @@ private object VisitCache(ServiceCallSite callSite, RuntimeResolverContext conte
             }
         }
 
+        private object ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
+        {
+            Dictionary<ServiceCacheKey, object> resolvedServices = serviceProviderEngine.ResolvedServices;
```

Comment:
Another issue pointed out by @pakrym is that we're not synchronizing access to the dictionary, so multiple resolves can happen concurrently now and we'll try to add them to the dictionary concurrently without a lock. We may want to change the singleton dictionary to be a concurrent dictionary.

Reply:
Actually a simple lock will suffice.

Reply:
but if they have the same type they share a lock. is that not enough?

Reply:
Consider 2 different threads resolving 2 different singletons at the same time. They'll each have their own lock and will resolve abs party on the same non concurrent dictionary